### PR TITLE
feat: Add SessionBatchSampler for grouping samples into per-session batches

### PR DESCRIPTION
**Pull request by AlexandreAndr**
## Summary

Adds `SessionBatchSampler`, a `Sampler` wrapper that groups any inner sampler's `DatasetIndex` output into per-session batches (no batch mixes sessions), with optional batch shuffling and `drop_last` control. Extracted from the NDT multi-session notebook, where it was previously defined inline.

- New: `torch_brain/samplers/session_batch_sampler.py`, registered in `torch_brain/samplers/__init__.py`
- Tests: `tests/test_sampler.py::test_session_batch_sampler` (per-session grouping, `drop_last` True/False, invalid `batch_size`, `len()` consistency)

### CODE REVIEW

File: torch_brain/samplers/session_batch_sampler.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        ... )
+        >>> sampler = SessionBatchSampler(inner_sampler, batch_size=16)
+    """
+
+    def __init__(
+        self,
+        sampler: torch.utils.data.Sampler,
+        batch_size: int,
+        *,
+        shuffle_batches: bool = False,
```
Comment (divyansha1115) [inline torch_brain/samplers/session_batch_sampler.py]: any specific reason to why default shuffle is false? usually default is set to true.

**Review by divyansha1115 — COMMENTED** [review]
can you update the change log

**Comment by AlexandreAndr** [comment]
Hi, for the changelog I think we are trying to get rid of it by having more informative PR names

File: torch_brain/samplers/session_batch_sampler.py (hunk 1 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: That is a good point, we might consider it true by default yes. To keep in mind the sampler that wraps arround it has it's own internal logic for the sampling but I agree that having it by default set to true might be better

Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: just updated it

File: torch_brain/samplers/session_batch_sampler.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        >>> inner_sampler = RandomFixedWindowSampler(
+        ...     sampling_intervals=sampling_intervals,
+        ...     window_length=1.0,
+        ... )
+        >>> sampler = SessionBatchSampler(inner_sampler, batch_size=16)
+    """
+
+    def __init__(
+        self,
+        sampler: torch.utils.data.Sampler,
```
Comment (vinamarora8) [inline torch_brain/samplers/session_batch_sampler.py]:
```suggestion
        sampler: torch.utils.data.Sampler[DatasetIndex],
```

I think this will work?

File: torch_brain/samplers/session_batch_sampler.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+        >>> sampling_intervals = {
+        ...     "session_1": Interval(0.0, 100.0),
+        ...     "session_2": Interval(0.0, 200.0),
+        ... }
+        >>> inner_sampler = RandomFixedWindowSampler(
+        ...     sampling_intervals=sampling_intervals,
+        ...     window_length=1.0,
+        ... )
+        >>> sampler = SessionBatchSampler(inner_sampler, batch_size=16)
```
Comment (vinamarora8) [inline torch_brain/samplers/session_batch_sampler.py]: Would be amazing if u also add small snipped for how to use a batch sampler with the torch dataloader!

File: torch_brain/samplers/session_batch_sampler.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+from collections.abc import Iterator
+
+import torch
+
+from torch_brain.datasets import DatasetIndex
+
+
+class SessionBatchSampler(torch.utils.data.Sampler[list[DatasetIndex]]):
+    r"""Wraps any sampler yielding :class:`~torch_brain.datasets.DatasetIndex` and
```
Comment (vinamarora8) [inline torch_brain/samplers/session_batch_sampler.py]: Another idea for the name: `SessionWiseBatchSampler`, maybe?

File: torch_brain/samplers/session_batch_sampler.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        self._prepare_cache()
+        return len(self._batches_cache)
+
+    def __iter__(self) -> Iterator[list[DatasetIndex]]:
+        r"""Returns an iterator over per-session batches of
+        :class:`~torch_brain.datasets.DatasetIndex`."""
+        self._prepare_cache()
+        batches = self._batches_cache
+        self._batches_cache = None  # reset so next epoch rebuilds
+        return self._yield_batches(batches)
```
Comment (vinamarora8) [inline torch_brain/samplers/session_batch_sampler.py:115]:
```suggestion
    def __iter__(self) -> Iterator[list[DatasetIndex]]:
        r"""Returns an iterator over per-session batches of
        :class:`~torch_brain.datasets.DatasetIndex`."""
        batches = self._build_batches()
        return self._yield_batches(batches)
```

Maybe this is simpler? Looks to me that you want to build batches every epoch, so there is no need for a separate `_prepare_cache` etc.. Apologies if I'm wrong!

File: torch_brain/samplers/session_batch_sampler.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        ... )
+        >>> sampler = SessionBatchSampler(inner_sampler, batch_size=16)
+    """
+
+    def __init__(
+        self,
+        sampler: torch.utils.data.Sampler,
+        batch_size: int,
+        *,
+        shuffle_batches: bool = True,
```
Comment (vinamarora8) [inline torch_brain/samplers/session_batch_sampler.py]:
```suggestion
        shuffle: bool = False,
```

I think `shuffle_batches` should simply be `shuffle`? And also it should default to `False` to stay consistent with other torch stuff (like how shuffle is False by default in `torch.utils.data.DataLoader`).

File: torch_brain/samplers/session_batch_sampler.py (hunk 3 above)
Comment (vinamarora8) [reply torch_brain/samplers/session_batch_sampler.py]:
```suggestion
        >>> inner_sampler = RandomFixedWindowSampler(
        ...     sampling_intervals=sampling_intervals,
        ...     window_length=1.0,
        ... )
        >>> batch_sampler = SessionBatchSampler(inner_sampler, batch_size=16)
        >>> loader = torch.utils.data.DataLoader(
        ...     dataset=your_dataset,
        ...     batch_sampler=batch_sampler,
        ... )
```

**Review by vinamarora8 — COMMENTED** [review]
A very clean way to make session-wise batches, without making a batch-samplers for each single-sample samplers! Left some comments below

File: torch_brain/samplers/session_batch_sampler.py (hunk 1 above)
Comment (vinamarora8) [reply torch_brain/samplers/session_batch_sampler.py]: Hmm.. torch's DataLoader's default value for shuffle is False. @divyansha1115 curious, where have you seen default shuffle being True?

https://docs.pytorch.org/docs/2.13/data.html

File: torch_brain/samplers/session_batch_sampler.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        r"""Returns the total number of batches across all sessions."""
+        self._prepare_cache()
+        return len(self._batches_cache)
+
+    def __iter__(self) -> Iterator[list[DatasetIndex]]:
+        r"""Returns an iterator over per-session batches of
+        :class:`~torch_brain.datasets.DatasetIndex`."""
+        self._prepare_cache()
+        batches = self._batches_cache
+        self._batches_cache = None  # reset so next epoch rebuilds
```
Comment (milosobral) [inline torch_brain/samplers/session_batch_sampler.py]: Do we always want to rebuild at every epoch? Should this maybe be an option?

File: torch_brain/samplers/session_batch_sampler.py (hunk 6 above)
Comment (milosobral) [reply torch_brain/samplers/session_batch_sampler.py]: I actually think that `shuffle_batches` is slightly better given that it makes it clear that we are shuffling the order of the batches, not the order of the samples inside of each batch

**Review by milosobral — COMMENTED** [review]
I have a small comment on resetting but otherwise looks great!

File: torch_brain/samplers/session_batch_sampler.py (hunk 2 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: yes good point I'll also update another sampler typing to follow this

File: torch_brain/samplers/session_batch_sampler.py (hunk 3 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: Done!

File: torch_brain/samplers/session_batch_sampler.py (hunk 7 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: It is a good point!

Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: 1) We want to rebuild at each epoch for the random sampler. Otherwise, you would have always the same samples within batches (and batches sequence would be random or not)
2) I added a `cache_batches` args that is `True` by default as shuffle is `False`

File: torch_brain/samplers/session_batch_sampler.py (hunk 5 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py:115]: Hi, I agree it would be more elegant and I am not a fan of all the subfunctions. 
The need for the `_prepare_cache` comes from the fact that you probably call `__len()__` before `__iter__` and you want the cached from both to be aligned

File: torch_brain/samplers/session_batch_sampler.py (hunk 4 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: I am neutral to it, @milosobral @divyansha1115 do you have a preference?

File: torch_brain/samplers/session_batch_sampler.py (hunk 6 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: I also like something more descriptive. For the false and true default behavior, I will post something bellow!

File: torch_brain/samplers/session_batch_sampler.py (hunk 4 above)
Comment (divyansha1115) [reply torch_brain/samplers/session_batch_sampler.py]: I would have slight preference towards SessionWiseBatchSampler. The initial one sound a little off to me.

File: torch_brain/samplers/session_batch_sampler.py (hunk 6 above)
Comment (divyansha1115) [reply torch_brain/samplers/session_batch_sampler.py]: I think shuffle would be better here. Given we call it batch sampler it would be clear that the shuffle is on batches. 
Also all sampler having somewhat similar and consistent arguments is easier to remember and configure.

I also agree on the default being changed to false. I was interpreting the sampler in a different way before, but looking deeper I think False as default is more appropriate.

File: torch_brain/samplers/session_batch_sampler.py (hunk 4 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: okay let move to it!

File: torch_brain/samplers/session_batch_sampler.py (hunk 6 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: For concistency of the samplers I move it to `false`

Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: and move to `shuffle`

File: torch_brain/samplers/session_batch_sampler.py (hunk 7 above)
Comment (AlexandreAndr) [reply torch_brain/samplers/session_batch_sampler.py]: For 2) cc @divyansha1115 @vinamarora8 , What do you think of this arg?
For me the tradeoff is that caching batch can save time for the deterministic samplers but complexify slightly the interface
